### DISCUSSION TITLE
why do you need a broker?

### DISCUSSION THREAD

**Discussion by martin3000** [Koenkk/zigbee2mqtt#10515] [General]
When I use zigbee2mqtt with an application, lets say home assistant, why do I need a broker? Isn't z2m just sending text messages over the network and can't the messages be received by an application, without a routing broker in between?

**Comment by mausch** [comment]
If the question is why MQTT requires a broker: MQTT is a pubsub protocol, pretty much all of them require a broker by design to route/filter messages etc. See https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern#Topologies

If the question is why zigbee2mqtt requires MQTT, well, it's in its name :slightly_smiling_face: 
MQTT is a generic standard protocol in IoT, which buys you modularity and flexibility.
For example:
- you can choose the broker implementation depending on your needs e.g. there's https://www.hivemq.com/ , https://mosquitto.org/ , etc.
- you could publish messages to Home Assistant from other sources not just Z2M without needing to change anything in Home Assistant.
- you could hook up Z2M to other consumers instead of / in addition to Home Assistant e.g. https://nodered.org/

There are also other solutions that connect zigbee to Home Assistant without MQTT e.g. https://www.home-assistant.io/integrations/zha/

**Reply by martin3000** [reply]
the question is when there is only one publisher and only one subscriber, why can't the subscriber not directly listen to the publisher.

**Reply by mausch** [reply]
MQTT is a client-server protocol. Both Home Assistant and zigbee2mqtt are clients to the broker. If there were no broker then Home Assistant would just have to implement/embed a MQTT server/broker.
